File: q4/q4bSemaphoreImpl.java
import java.util.Random;
import java.util.concurrent.Semaphore;

public class Main {

    public static void main(String args[]) {
        System.out.println("gateway=" + gateway(5));
    }

    public static int gateway(int numReplicas) {
        Sum sum = new Sum();
        Thread[] threads = new Thread[numReplicas];
        Request request = new Request(sum);
        for (int i = 0; i < numReplicas; i++) {
            Thread thread = new Thread(request, "request" + i);
            threads[i] = thread;
        }
        
        Semaphore semaphore = new Semaphore(0);
        Joiner joiner = new Joiner(threads, semaphore);
        Thread joinerThread = new Thread(joiner, "joiner");
        Timer timer = new Timer(numReplicas, semaphore, sum);
        Thread timerThread = new Thread(timer, "timer");
        
        timerThread.start();        
        for(Thread thread: threads){
            thread.start();
        }
        joinerThread.start();

        try {
            semaphore.acquire();      
        } catch (InterruptedException e) {}
        return sum.getValue();
    }
    
    public static class Joiner implements Runnable {
        private Thread[] threads;
        private Semaphore semaphore;

        public Joiner (Thread[] threads, Semaphore semaphore) {
            this.threads = threads;
            this.semaphore = semaphore;
        }

        public void run() {
            try {
                System.out.println("Joiner waititng...");
                for (int i = 0; i < this.threads.length; i++) {                
                    this.threads[i].join();
                    System.out.println("Thread " + this.threads[i].getName() + " finished.");
                }
            } catch (InterruptedException e) {}
            this.semaphore.release();
        }
    }

    public static class Timer implements Runnable {
        private Semaphore semaphore;
        private int threadNumber;
        private Sum sum;

        public Timer(int threadNumber, Semaphore semaphore, Sum sum) {
            this.threadNumber = threadNumber;
            this.semaphore = semaphore;
            this.sum = sum;
        }

        @Override
        public void run() {
            try {
                System.out.println("Timer will sleep 16 seconds.");
                Thread.sleep(16 * 1000); // Thread.sleep sleeps milliseconds
            } catch (InterruptedException e) {}
            if (this.sum.getAddsCount() < this.threadNumber) {
              this.sum.setTimeout();
            }
            this.semaphore.release();
        }
    }

    public static class Request implements Runnable {
        private Sum sum;

        public Request (Sum sum) {
            this.sum = sum; 
        }

        public void run() {
            int randomNumber = (new Random()).nextInt(30) + 1; // Obtain a number between [1 - 30].
            try {
                System.out.println("Request will sleep " + randomNumber + " seconds.");
                Thread.sleep(randomNumber * 1000); // Thread.sleep sleeps milliseconds
            } catch (InterruptedException e) {}
            this.sum.add(randomNumber);
        }
    }

    public static class Sum {
        private int value, addsCount;
        private boolean timeout;

        public synchronized void add(int value) {
            if (!this.timeout) {
                this.addsCount++;
                this.value += value;
            }
        }

        public synchronized void setTimeout() {
            this.timeout = true;
            this.value = -1;
        }

        public synchronized int getAddsCount() {
            return this.addsCount;
        }

        public synchronized int getValue() {
            return this.value;
        }
    }  

}

File: q4/q4b-semaphore-impl.c
#include <stdio.h>
#include <assert.h>
#include <pthread.h>
#include <semaphore.h>
#include <time.h>

sem_t semaphore;
pthread_mutex_t mutex;
pthread_t pthreads[5];
int numberOfThreads = 5;
int threadsFinished = 0;
int sum = 0;

void *joiner (void *args) {
    int joiner_sum = 0;
    printf("Joiner waiting...\n");
    for (int i = 0; i < numberOfThreads; i++) {
        int aux = 0;
        pthread_join(pthreads[i], &aux);    
        printf("Thread %d finished.\n", i);
        joiner_sum += aux;        
    }

    pthread_mutex_lock(&mutex);
    if (sum != -1) {
        sum = joiner_sum;
    }
    sem_post(&semaphore);
    pthread_mutex_unlock(&mutex);
    
    pthread_exit(NULL);
}

void *timer (void *args) {
    printf("Timer will sleep 16 seconds.\n");
    sleep(16); // Sleeps seconds

    pthread_mutex_lock(&mutex);
    if (threadsFinished < numberOfThreads) {
        sum = -1;
    }
    sem_post(&semaphore);
    pthread_mutex_unlock(&mutex);

    pthread_exit(NULL);
}

void *request (void *args) {
    int random_number = (rand() % 30) + 1; // // Obtain a number between [1 - 30].
    printf("Request will sleep %d seconds\n", random_number);
    sleep(random_number); // Sleeps seconds  
    
    pthread_mutex_lock(&mutex);
    threadsFinished++; 
    pthread_mutex_unlock(&mutex);  
    
    pthread_exit(random_number);
}

int gateway (int num_replicas) {
    for (int i = 0; i < num_replicas; i++) {
        pthread_create (&pthreads[i], NULL, &request, (void*) i);
    }

    pthread_t joinerThread;
    pthread_create(&joinerThread, NULL, &joiner, NULL);
    pthread_t timerThread;
    pthread_create(&timerThread, NULL, &timer, NULL);

    sem_wait(&semaphore);
    pthread_mutex_lock(&mutex);
    int sumValue = sum;
    pthread_mutex_unlock(&mutex);
    return sumValue;
}

int main (int argc, char *argv[]) {
    srand ( time(NULL) );
    sem_init(&semaphore, 0, 0);
    pthread_mutex_init(&mutex, NULL);
    printf("gateway=%d\n", gateway(numberOfThreads));
}